pennylane/transforms/decompose.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+A transform for decomposing quantum circuits into user defined gate sets. Offers an alternative to the more device-focused decompose transform.
+"""
+# pylint: disable=protected-access, too-many-arguments
+
+
+import pennylane as qml
+from pennylane.operation import StatePrepBase
+from pennylane.tape import QuantumScript
+from pennylane.transforms.core import transform
+
+
+def null_postprocessing(results):
+    """A postprocessing function returned by a transform that only converts the batch of results
+    into a result for a single ``QuantumTape``.
+    """
+    return results[0]
+
+
+@transform
+def decompose(tape, gate_set=None, max_expansion=None):
+    """Decomposes operations found in a quantum circuit into a desired gate set.
+
+    Args:
+        tape (QuantumScript or QNode or Callable): a quantum circuit.
+        gate_set (Callable or set, optional): A set of decomposition gates or a condition they satisfy. Defaults to None.
+        max_expansion (int, optional): The maximum depth of the expansion. Defaults to None.
+
+    Returns:
+        qnode (QNode) or quantum function (Callable) or tuple[List[QuantumScript], function]:
+
+        The decomposed circuit. The output type is explained in :func:`qml.transform <pennylane.transform>`.
+
+    Raises:
+        Exception: Type defaults to ``qml.DeviceError``.
+            Raised if an operator is not accepted and does not define a decomposition, or if
+            the decomposition enters an infinite loop and raises a ``RecursionError``.
```

Comment:
For the "user friendly" version, we probably want to consider the options:

1) Just leaving the non-decomposable operation as is
2) Leaving the non-decomposable operation in, but warning that it didn't get to the target gateset
3) Raising a `DecompositionUndefinedError`

For the device version, we needed to raise an error and halt execution before sending something to the device that it didn't support.  If the main purpose of this transform is for visualization, inspection, and user compilation, then the cost of leaving something in the circuit outside the gateset is a lot lower.

I'm thinking I'm preferring option 1 at the moment.